File: agent-orchestration-protocol/scripts/aop-file-watcher.js
#!/usr/bin/env node
// ============================================================
// aop-file-watcher.js — Node.js file watcher using chokidar
// ============================================================
//
// Purpose:  Monitors a directory for AOP_COMPLETE_*.json creation
//           events and writes signal files for the Orchestrator.
//           Detection latency: <1 second.
//
// Usage:    node scripts/aop-file-watcher.js <watch_dir> <signal_dir>
//
// Dependencies: chokidar (npm install chokidar)
//
// Example:
//   node scripts/aop-file-watcher.js /c/ai/target-project /c/ai/target-project/signals
//
// Signal files:
//   When AOP_COMPLETE_foo_a1b2.json is created, the watcher writes:
//     <signal_dir>/SIGNAL_AOP_COMPLETE_foo_a1b2.txt
//   containing the full path to the artifact.
// ============================================================

const fs = require('fs');
const path = require('path');

let chokidar;
try {
  chokidar = require('chokidar');
} catch (e) {
  console.error('ERROR: chokidar not installed. Run: npm install chokidar');
  process.exit(1);
}

const [watchDir, signalDir] = process.argv.slice(2);

if (!watchDir || !signalDir) {
  console.error(`Usage: ${process.argv[1]} <watch_dir> <signal_dir>`);
  process.exit(1);
}

// Ensure signal directory exists
fs.mkdirSync(signalDir, { recursive: true });

chokidar
  .watch(path.join(watchDir, 'AOP_COMPLETE_*.json'), { ignoreInitial: true })
  .on('add', (filePath) => {
    const stem = path.basename(filePath, '.json');
    const sig = path.join(signalDir, `SIGNAL_${stem}.txt`);
    fs.writeFileSync(sig, filePath);
    console.log(`[watcher] Signal written: SIGNAL_${stem}.txt`);
  });

console.log(`[watcher] Monitoring ${watchDir} for AOP_COMPLETE_*.json ...`);

File: agent-orchestration-protocol/scripts/aop-file-watcher.py
#!/usr/bin/env python3
# ============================================================
# aop-file-watcher.py — Python file watcher using watchdog
# ============================================================
#
# Purpose:  Monitors a directory for AOP_COMPLETE_*.json creation
#           events and writes signal files for the Orchestrator.
#           Detection latency: <1 second.
#
# Usage:    python3 scripts/aop-file-watcher.py <watch_dir> <signal_dir>
#
# Dependencies: watchdog (pip install watchdog)
#
# Example:
#   python3 scripts/aop-file-watcher.py /c/ai/target-project /c/ai/target-project/signals
#
# Signal files:
#   When AOP_COMPLETE_foo_a1b2.json is created, the watcher writes:
#     <signal_dir>/SIGNAL_AOP_COMPLETE_foo_a1b2.txt
#   containing the full path to the artifact.
#
# The Orchestrator checks for signal files instead of polling:
#   SIGNAL="SIGNAL_AOP_COMPLETE_${tid}_${SESSION_ID}.txt"
#   if test -f "${SIGNAL_DIR}/${SIGNAL}"; then
#     ARTIFACT_PATH=$(cat "${SIGNAL_DIR}/${SIGNAL}")
#     EXECUTOR_STATUS[$tid]="COMPLETE"
#   fi
# ============================================================

import sys
import time
from pathlib import Path

try:
    from watchdog.observers import Observer
    from watchdog.events import FileSystemEventHandler
except ImportError:
    print("ERROR: watchdog not installed. Run: pip install watchdog", file=sys.stderr)
    sys.exit(1)

if len(sys.argv) != 3:
    print(f"Usage: {sys.argv[0]} <watch_dir> <signal_dir>", file=sys.stderr)
    sys.exit(1)

WATCH_DIR = sys.argv[1]
SIGNAL_DIR = sys.argv[2]

# Ensure signal directory exists
Path(SIGNAL_DIR).mkdir(parents=True, exist_ok=True)


class ArtifactHandler(FileSystemEventHandler):
    def on_created(self, event):
        p = Path(event.src_path)
        if p.name.startswith("AOP_COMPLETE_") and p.suffix == ".json":
            sig = Path(SIGNAL_DIR) / f"SIGNAL_{p.stem}.txt"
            sig.write_text(str(p))
            print(f"[watcher] Signal written: {sig.name}", flush=True)


observer = Observer()
observer.schedule(ArtifactHandler(), WATCH_DIR, recursive=False)
observer.start()
print(f"[watcher] Monitoring {WATCH_DIR} for AOP_COMPLETE_*.json ...", flush=True)

try:
    while True:
        time.sleep(1)
except KeyboardInterrupt:
    observer.stop()

observer.join()
